### PR TITLE
fix(url parser): preserve auth creds when composing conn string 

### PR DESCRIPTION
**Pull request by jlord**
Preserve the auth credentials used with a dns/txt record uri when composing a new connection string for the url parser. 

This fix will also be backported to 2.2 which got support for dns/txt records. 

NODE-1286

### CODE REVIEW

File: lib/url_parser.js [hunk 1]
```diff
@@ -49,7 +49,8 @@ module.exports = function(url, options, callback) {
     }
 
     let connectionStrings = addresses.map(function(address, i) {
-      if (i === 0) return `mongodb://${address.name}:${address.port}`;
+      let base = result.auth ? `mongodb://${result.auth}@` : `mongodb://`;
```
Comment (daprahamian) [inline lib/url_parser.js]: Can you compute `base` outside of the loop so that you don't do it on every iteration?

**Comment by IrishAfrican** [comment]
This still doesn't seem to work correctly for me.
If you have a URIEncoded Password in a connection string like such:  

> mongodb+srv://test-user:3#f!69A8AZu9@cluster0-abcde.mongodb.net/mytestdb

becoming:

> mongodb+srv://test-user:3%23f!69A8AZu9@cluster0-abcde.mongodb.net/mytestdb

when **parser.parse** returns the parsed object it decodes the password to be  '3#f!69A8AZu9',
but the code later that tries to decode and re-encode the password (to determine if the pwd was encoded correctly) fails as the password had already been decoded by the parser.parse method.

Note that the new NodeJS URL API whereby one says:
```
let myUrl = URL(the_url)
myUrl.password
```
behaves differently and does not decode the password

**Comment by jlord** [comment]
@IrishAfrican thanks so much for reporting this! I've made a ticket for us in JIRA and we'll make the fix 👍  

https://jira.mongodb.org/browse/NODE-1309

**Comment by IrishAfrican** [comment]
Thanks Jessica.

On a different but related issue, the latest driver may have a problem with the case whereby the auth database and the database being connected to, no longer works correctly, or at least not from 3rd party systems such as Mongoose.

The MongoDB Atlas, says that even on a v3.6 compliant driver I should use the **/mytestdb** syntax to set my database name even though I authenticate against the admin database as shown in my example previously
> mongodb+srv://test-user:3#f!69A8AZu9@cluster0-abcde.mongodb.net/mytestdb 

In the case of [Mongoose](http://mongoosejs.com/) they use the MongoDB driver library url_parser to parse the URL and get the DB name. In the latest version of the driver this seems to return 'admin'.

If I use the 3.4 syntax as shown below it seems to work fine:
> mongodb://test-user:3#f!69A8AZu9@cluster0-shard-00-00-abcde.mongodb.net:27017,cluster0-shard-00-01-abcde.mongodb.net:27017,cluster0-shard-00-02-abcde.mongodb.net:27017/mytestdb?ssl=true&replicaSet=cluster0-shard-0&authSource=admin

I am not sure if you would consider this a bug as it does not seem that the URL Parser is supposed to be a public API, but the url_parser does seem to have code that tries to set **dbName** to something other than its default of 'admin' in certain cases, and it works for the 3.4 syntax so I am assuming it is a bug.

Obviously, if you feel that this is not a bug and Mongoose should not be using the MongoDB parser I will log an issue on their side.

Thanks

**Comment by mbroadst** [comment]
@IrishAfrican please file a ticket on JIRA to discuss this, thanks!

**Comment by IrishAfrican** [comment]
Cool, have done: [NODE-1321](https://jira.mongodb.org/browse/NODE-1321)

**Pull request merged**